packages/opencensus-core/src/exporters/types.ts

```diff
@@ -25,7 +25,8 @@ export interface Exporter extends modelTypes.SpanEventListener {
    * Sends a list of root spans to the service.
    * @param rootSpans A list of root spans to publish.
    */
-  publish(rootSpans: modelTypes.RootSpan[]): void;
+  // tslint:disable-next-line:no-any
+  publish(rootSpans: modelTypes.RootSpan[]): Promise<any>|void;
```

Comment:
Two notes:
1. What is the meaning of the deferred value being returned here? How are users expected to treat this value? If there are no good answers to this question it should just be `Promise<void>`.
2. Returning either a Promise or `void` complicates handling the return value. Can we only return `Promise`?

Reply:
The ideia is that by returning `Promise<any> | void` we account for `publish` methods that returns a Promise, while giving freedom to future collaborators to make a method that returns `void`. Currently, `NoopExporter` and `ConsoleExporter` return void, but if you prefer, I can change them to return `Promise<void>` and restrict the interface, no problems with that IMO.

Reply:
It's OK for `publish` to inherently not contain anything asynchronous, but it would still be more consistent to return a `Promise.resolve()` instead of `void` anyway. If we consider the code that will call this function:
```js
const result = exporter.publish();
if (result) {
  result.then(next);
} else {
  next(); // note: zalgo alert
}
```
Why even have this branching logic, when we could just do
```js
publish().then(next);
```

That was the original context behind the second question. If you could make the change in `*Exporter` that would be great.

Reply:
Hmm, got it. Made the change but had to make `Promise<number|string|void>` to account for some exporter implementations that resolve `number` or `string`.

Reply:
`Promise<number|string|void>` goes back to my original first question. Do we use the return value in a meaningful way? If so, it would be better to standardize now what that return value is (maybe a boolean or status code). Otherwise, if we don't use it, we should just make it `void`.

BTW, the answer to this question might not exist today. It's hard to know how to standardize the return value if we are currently not using it. So I can merge this PR, just something to think about.

Reply:
I could see the return value being used in tests only, but not in applications. I'm not sure if we should standardize it tho, goes back to that thing about future contributors. But I understand your point of view. Would be great to discuss that latter.